### ISSUE TITLE
Early Proposal: Action Use Policy

### ISSUE THREAD

**Issue by ghost** [ossf/allstar#241]
The Action Use policy would allow orgs to specify rules for the use of Actions in their repos.

This policy should only be configured at the org level so that org security managers can oversee the use of Actions.

Here is the latest config concept I have for the policy:
```yaml
rules:
  # Require: Require Action confirming go.mod implies go.sum on all Go repos
  # the action requires that a go.sum be present if there is a go.mod
  - name: "Require Action confirming go.mod implies go.sum on all Go repos"
    repo:
      name: *
      lang:
        - go
    method: require
    mustPass: true
    actions:
      - name: "helpful/require-go-sum"
        version: ">= 2.0.2"
  # Require: Require a passing DAST Action for all repos
  - name: "Require a passing DAST Action for all repos"
    repo:
      name: *
    method: require
    # default count is number of listed Actions
    # in this case, only one of the listed actions is required to pass
    count: 1
    mustPass: true
    actions:
      - name: "stackhawk/hawkscan-action"
        version: ">= 2.0.0"
      - name: "projectdiscovery/nuclei-action"
        # Non-semver versions will check for a branch name / commit hash
        version: "main"
  # Allow: Allow trusted Actions
  # actions/checkout and actions/cache@0865c47f36e68161719c5b124609996bb5c40129 allowed for all repos
  - name: "Allow trusted Actions"
    method: allow
    actions:
      - name: "actions/*"
      - name: "actions/cache"
        version: "0865c47f36e68161719c5b124609996bb5c40129"
  # Deny: Deny default
  # deny by default (allowlist strategy)
  - name: "Deny default"
    method: deny
```

The policy should give __very__ detailed errors when it fails. Things like:

```
Action helpful/require-go-sum version 1.0.3:
-> does not meet version requirement ">= 2.0.2" for require rule "Require Action confirming go.mod implies go.sum on all Go repos"
-> is not listed in require rule "Require a passing DAST Action for all repos"
-> is not listed in allow rule "Allow trusted Actions"
-> denied by deny rule "Deny default"
```

Feedback on the concept would be appreciated! Some questions to consider:
- Do you think the config is too verbose or confusing?
- Should there be a simpler way to set defaults than using rule precedence or is this reasonable?
- Can you think of other ways to specify repo targets that might be useful?

**Comment by ghost** [comment]
I wonder if it would make sense to add a "remote rules" feature, where rules from another repo / file could be imported as part of a step (limited to some number in a config, including recursive imports), so that they can be organized differently and Allstar can have some importable sample / base allowlist configs.

**Comment by ghost** [comment]
Allow Action name wildcards. The example rules have been updated.

```yaml
- name: "Allow trusted Actions"
  method: allow
  actions:
    - name: "actions/*"
    - name: "ossf/*"
```

**Comment by ghost** [comment]
Closed, formal proposal created in #246
